## IV Building model

## 1. Import libraies and data set

In [18]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import train_test_split

In [2]:
app_train = pd.read_csv('application_train_data.csv')
app_test = pd.read_csv('application_test_data.csv')
id = pd.read_csv('id_submit.csv')

In [3]:
X_train = app_train.drop(columns = 'TARGET')
X_test = app_test
y_train = app_train['TARGET']

In [4]:
del app_train
del app_test

## 2. Using gridsearch to find the best params

In [ ]:
# pipeline = Pipeline([
#     ('scaler', StandardScaler()),
#     ('model', LogisticRegression(random_state=42))
# ])

# param_grid = {
#     'model__C': [0.01, 0.1, 1, 10, 100], 
#     'model__solver': ['lbfgs', 'liblinear'],  
#     'model__penalty': ['l2'],  
#     'model__max_iter': [100, 200, 500], 
# }

# grid_search = GridSearchCV(
#     estimator=pipeline,
#     param_grid=param_grid,
#     cv=5,  
#     scoring='roc_auc',  
#     n_jobs=-1,  
#     verbose=2  
# )

In [ ]:
# grid_search.fit(X_train_split, y_train_split)

# print("Best Parameters:", grid_search.best_params_)
# print(f"Best Cross-Validation ROC AUC: {grid_search.best_score_:.4f}")

## 3. Test the model on the training set using validation

In [5]:
model = LogisticRegression(C = 10, max_iter = 1000,solver = 'liblinear')

In [20]:
# Splitting the training data into train and validation sets
X_train_split, X_valid, y_train_split, y_valid = train_test_split(X_train, y_train, test_size=0.2, random_state=42, stratify=y_train)

In [22]:
scaler = StandardScaler()

# Fit the scaler on the training data and transform both training and validation data
X_train_scaled = scaler.fit_transform(X_train_split)
X_valid_scaled = scaler.transform(X_valid)

In [23]:
# Train and evaluate models
model.fit(X_train_scaled, y_train_split)
y_pred_valid = model.predict_proba(X_valid_scaled)[:, 1]
roc_auc = roc_auc_score(y_valid, y_pred_valid)
print(f"ROC AUC: {roc_auc:.4f}")

ROC AUC: 0.7729


### With logistic regression we have achieve the roc auc: 0.7729

## 4. Train the whole data for submission

In [ ]:
# scaler = StandardScaler()

# # Fit the scaler on the training data and transform both training and validation data
# X_train_scaled = scaler.fit_transform(X_train_split)
# X_valid_scaled = scaler.transform(X_valid)

In [ ]:
# model.fit(X_train_scaled, y_train)
# y_test_pred = model.predict_proba(X_test_scaled)[:, 1]
# submission = pd.DataFrame({'SK_ID_CURR': id['SK_ID_CURR'], 'TARGET': y_test_pred})

In [ ]:
# submission.shape

In [ ]:
# submission.head()

In [ ]:
# submission.to_csv("submission.csv", index=False)